docs/proposals/kubectl-login.md

```diff
@@ -0,0 +1,233 @@
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Kubectl Login Subcommand
+
+**Authors**: Eric Chiang (@ericchiang)
+
+## Goals
+
+`kubectl login` attempts to guide the user's experience when configuring
+credentials for kubectl. It should assist users who are first experiencing
+Kubernetes, users who are partially setup, and help users who are already
+prepared debug their own credentials.
+
+## Design
+
+The "login" subcommand helps users move towards a fully functional kubeconfig by
+evaluating the current state of the kubeconfig and trying to prompt the user for
+and validate the necessary information to login to the kubernetes cluster.
+
+This is inspired by a similar tools such as:
+
+ * [os login](https://docs.openshift.org/latest/cli_reference/get_started_cli.html#basic-setup-and-login)
+ * [gcloud auth login](https://cloud.google.com/sdk/gcloud/reference/auth/login)
+ * [aws configure](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html)
+
+The steps taken are:
+
+1. If no cluster configured, prompt user for cluster information.
+2. If no user is configured, discover the authentication strategies supported by the API Server.
+3. Prompt the user for some information based on the authentication strategy they choose.
+4. Attempt to login as a user, including authentication challenges such as OAuth2 flows, and display user info.
+
+Importantly, each of these steps only trigger if kubeconfig isn't configured
+(or is improperly configured). Users with fully configured kubeconfigs will only
+see the user they're logged in as, useful for opaque credentials such as X509
+certs or bearer tokens.
+
+The command differs from `kubectl config set-credentials` by:
+
+* Communicating with the API Server to determine if the user is supplying valid auth events.
+* Validating input and being opinionated about the input it asks for.
+* Triggering authentication challenges for example:
+  * Basic auth: Actually try to communicate with the API server.
+  * OpenID Connect: Create an OAuth2 redirect.
+
+
+## Server endpoints
+
+On the API Server side this command requires two non-resource paths.  First a
+path to discover the authentication strategies supported by the API Server.
+
+> GET /auth/authenticators
+
+```
+{
+  "kind": "AuthentactionModes",
+  "authenticators": [
+    {
+      "mode": "tokenfile"
+    },
+    {
+      "mode": "oidc",
+      "config": {
+        "issuer": "https://accounts.google.com",
+        "trusted-client-id": "foobar"
+      }
+    },
+    {
+      "mode": "keystone",
+      "config": {
+        "url": "https://keystone.example.com"
+      }
+    },
+  ]
+}
+```
+
+And another path to get the user a request is logged in as.
+
+> GET /auth/whoami
+
+```
+{
+  "kind": "UserInfo",
+  "name": "janedoe",
+  "groups": [
+    "admins",
+    "kube-system"
+  ]
+}
+```
+
+To support these endpoints, add two top level types to `pkg/api/unversioned`.
```

Comment:
I don't want these to be unversioned.  I don't have confidence in us getting them right for all eternity.  I think I'd rather have versioned types and endpoints.